File: packages/hydra-java/src/main/java/hydra/sources/java/Environment.java
package hydra.sources.java;
import hydra.dsl.Types;
import static hydra.dsl.meta.Defs.unqualifiedDeps;
import hydra.packaging.Definition;
import hydra.packaging.Module;
import hydra.packaging.ModuleName;
import hydra.packaging.ModuleDependency;
import hydra.util.Maybe;

import java.util.Arrays;
import java.util.List;

import static hydra.dsl.java.Helpers.doc;
import static hydra.dsl.java.Helpers.typeDef;
import static hydra.dsl.java.Helpers.typeref;

/**
 * Environment types for Java code generation.
 *
 * <p>Mirror of
 * {@code packages/hydra-java/src/main/haskell/Hydra/Sources/Java/Environment.hs}.</p>
 */
public class Environment {
    public static final ModuleName NS = new ModuleName("hydra.java.environment");

    // Reference helpers, mirroring Haskell {@code environment}, {@code syntax}, etc.
    private static final ModuleName SYNTAX_NS = new ModuleName("hydra.java.syntax");
    private static final ModuleName CORE_NS = new ModuleName("hydra.core");
    private static final ModuleName GRAPH_NS = new ModuleName("hydra.graph");
    private static final ModuleName MODULE_NS = new ModuleName("hydra.packaging");
    private static final ModuleName TYPING_NS = new ModuleName("hydra.typing");

    private static hydra.core.Type env(String local) { return typeref(NS, local); }
    private static hydra.core.Type syntax(String local) { return typeref(SYNTAX_NS, local); }
    private static hydra.core.Type core(String local) { return typeref(CORE_NS, local); }
    private static hydra.core.Type graph(String local) { return typeref(GRAPH_NS, local); }
    private static hydra.core.Type modul(String local) { return typeref(MODULE_NS, local); }
    private static hydra.core.Type typing(String local) { return typeref(TYPING_NS, local); }

    /** Classification of a Java symbol for code generation. */
    private static Definition javaSymbolClass() {
        return typeDef(NS, "JavaSymbolClass",
            doc("Classification of a Java symbol for code generation",
                Types.union(
                    Types.field("constant", doc("A constant value", Types.unit())),
                    Types.field("nullaryFunction", doc("A nullary function (no arguments)", Types.unit())),
                    Types.field("hoistedLambda", doc("A hoisted lambda wrapped in type lambdas. The Int is the number of curried lambda parameters.", Types.int32())),
                    Types.field("unaryFunction", doc("A unary function (single argument)", Types.unit())),
                    Types.field("localVariable", doc("A local variable", Types.unit())))));
    }

    /** Feature flags for the target Java version. */
    private static Definition javaFeatures() {
        return typeDef(NS, "JavaFeatures",
            doc("Feature flags for the target Java version",
                Types.record(
                    Types.field("supportsDiamondOperator",
                        doc("Whether the diamond operator (<>) is supported (Java 7+)", Types.boolean_())))));
    }

    /** Aliases and context for Java code generation. */
    private static Definition aliases() {
        return typeDef(NS, "Aliases",
            doc("Aliases and context for Java code generation",
                Types.record(
                    Types.field("currentNamespace", doc("Current module name context", modul("ModuleName"))),
                    Types.field("packages", doc("Maps module names to Java package names",
                        Types.map(modul("ModuleName"), syntax("PackageName")))),
                    Types.field("branchVars", doc("Variables bound in pattern matching branches",
                        Types.set(core("Name")))),
                    Types.field("recursiveVars", doc("Variables that are self-recursive",
                        Types.set(core("Name")))),
                    Types.field("inScopeTypeParams", doc("Type parameters that are in scope (from method-level type parameters)",
                        Types.set(core("Name")))),
                    Types.field("polymorphicLocals", doc("Local variables that have polymorphic types (declared with raw types)",
                        Types.set(core("Name")))),
                    Types.field("inScopeJavaVars", doc("All in-scope Java variable names (for avoiding lambda parameter shadowing)",
                        Types.set(core("Name")))),
                    Types.field("varRenames", doc("Variable renames for avoiding shadowing (maps Hydra name to Java name)",
                        Types.map(core("Name"), core("Name")))),
                    Types.field("lambdaVars", doc("Lambda-bound variables (including hoisted captures with qualified names)",
                        Types.set(core("Name")))),
                    Types.field("typeVarSubst", doc("Type variable substitution: maps fresh inference variable names to canonical scheme variable names",
                        Types.map(core("Name"), core("Name")))),
                    Types.field("trustedTypeVars", doc("Type variables that actually appear in the method's formal parameter types",
                        Types.set(core("Name")))),
                    Types.field("methodCodomain", doc("The enclosing method's codomain (return type), used for casting pair expressions",
                        Types.optional(core("Type")))),
                    Types.field("thunkedVars", doc("Variables that have been thunked (wrapped in Supplier) for lazy evaluation",
                        Types.set(core("Name")))))));
    }

    /** Environment for Java code generation. */
    private static Definition javaEnvironment() {
        return typeDef(NS, "JavaEnvironment",
            doc("Environment for Java code generation",
                Types.record(
                    Types.field("aliases", doc("Aliases and context state", env("Aliases"))),
                    Types.field("graph", doc("Graph context for type inference", graph("Graph"))))));
    }

    private static final List<Definition> DEFINITIONS = Arrays.asList(
        javaSymbolClass(),
        javaFeatures(),
        aliases(),
        javaEnvironment());

    private static final List<ModuleDependency> DEPENDENCIES = unqualifiedDeps(
        SYNTAX_NS, CORE_NS, GRAPH_NS, MODULE_NS, TYPING_NS);

    public static final Module module_ = new Module(
        Maybe.just("Environment types for Java code generation"),
        NS,
        DEPENDENCIES,
        DEFINITIONS);
}

File: packages/hydra-java/src/main/haskell/Hydra/Sources/Java/Environment.hs
-- | Environment types for Java code generation.
-- These types support the Java coder and are used to track code generation state.

module Hydra.Sources.Java.Environment where

-- Standard imports for type-level sources outside of the kernel
import           Hydra.Kernel
import           Hydra.Dsl.Annotations
import           Hydra.Dsl.Bootstrap
import           Hydra.Dsl.Types                           ((>:), (@@))
import qualified Hydra.Dsl.Types                           as T
import qualified Hydra.Sources.Kernel.Types.Core           as Core
-- Additional imports
import qualified Hydra.Sources.Kernel.Types.Graph as Graph
import qualified Hydra.Sources.Kernel.Types.Packaging as Module
import qualified Hydra.Sources.Kernel.Types.Typing as Typing
import qualified Hydra.Sources.Java.Syntax as JavaSyntax


ns :: ModuleName
ns = ModuleName "hydra.java.environment"

module_ :: Module
module_ = Module {
            moduleName = ns,
            moduleDefinitions = (DefinitionType <$> definitions),
            moduleDependencies = unqualifiedDep <$> [JavaSyntax.ns, Core.ns, Graph.ns, Module.ns, Typing.ns],
            moduleMetadata = descriptionMetadata (Just "Environment types for Java code generation")}
  where
    definitions = [
      javaSymbolClass,
      javaFeatures,
      aliases,
      javaEnvironment]

-- | Aliases and context for Java code generation.
-- Tracks namespace mapping, variable scoping, type parameters, and other state
-- needed during the encoding of Hydra terms to Java syntax.
aliases :: TypeDefinition
aliases = def "Aliases" $
  doc "Aliases and context for Java code generation" $
  T.record [
    "currentNamespace">:
      doc "Current module namespace context" $
      modul "ModuleName",
    "packages">:
      doc "Maps namespaces to Java package names" $
      T.map (modul "ModuleName") (syntax "PackageName"),
    "branchVars">:
      doc "Variables bound in pattern matching branches" $
      T.set (core "Name"),
    "recursiveVars">:
      doc "Variables that are self-recursive" $
      T.set (core "Name"),
    "inScopeTypeParams">:
      doc "Type parameters that are in scope (from method-level type parameters)" $
      T.set (core "Name"),
    "polymorphicLocals">:
      doc "Local variables that have polymorphic types (declared with raw types)" $
      T.set (core "Name"),
    "inScopeJavaVars">:
      doc "All in-scope Java variable names (for avoiding lambda parameter shadowing)" $
      T.set (core "Name"),
    "varRenames">:
      doc "Variable renames for avoiding shadowing (maps Hydra name to Java name)" $
      T.map (core "Name") (core "Name"),
    "lambdaVars">:
      doc "Lambda-bound variables (including hoisted captures with qualified names)" $
      T.set (core "Name"),
    "typeVarSubst">:
      doc "Type variable substitution: maps fresh inference variable names to canonical scheme variable names" $
      T.map (core "Name") (core "Name"),
    "trustedTypeVars">:
      doc "Type variables that actually appear in the method's formal parameter types" $
      T.set (core "Name"),
    "methodCodomain">:
      doc "The enclosing method's codomain (return type), used for casting pair expressions" $
      T.optional (core "Type"),
    "thunkedVars">:
      doc "Variables that have been thunked (wrapped in Supplier) for lazy evaluation" $
      T.set (core "Name")]

core :: String -> Type
core = typeref Core.ns

def :: String -> Type -> TypeDefinition
def = datatype ns

environment :: String -> Type
environment = typeref ns

graph :: String -> Type
graph = typeref Graph.ns

-- | Environment for Java code generation.
javaEnvironment :: TypeDefinition
javaEnvironment = def "JavaEnvironment" $
  doc "Environment for Java code generation" $
  T.record [
    "aliases">:
      doc "Aliases and context state" $
      environment "Aliases",
    "graph">:
      doc "Graph context for type inference" $
      graph "Graph"]

-- | Feature flags for the target Java version.
javaFeatures :: TypeDefinition
javaFeatures = def "JavaFeatures" $
  doc "Feature flags for the target Java version" $
  T.record [
    "supportsDiamondOperator">:
      doc "Whether the diamond operator (<>) is supported (Java 7+)" $
      T.boolean]

-- | Classification of a Java symbol for code generation.
javaSymbolClass :: TypeDefinition
javaSymbolClass = def "JavaSymbolClass" $
  doc "Classification of a Java symbol for code generation" $
  T.union [
    "constant">:
      doc "A constant value" $
      T.unit,
    "nullaryFunction">:
      doc "A nullary function (no arguments)" $
      T.unit,
    "hoistedLambda">:
      doc "A hoisted lambda wrapped in type lambdas. The Int is the number of curried lambda parameters." $
      T.int32,
    "unaryFunction">:
      doc "A unary function (single argument)" $
      T.unit,
    "localVariable">:
      doc "A local variable" $
      T.unit]

modul :: String -> Type
modul = typeref Module.ns

syntax :: String -> Type
syntax = typeref JavaSyntax.ns

typing :: String -> Type
typing = typeref Typing.ns
